src/darwin/Framework/CHIP/MTROTAUnsolicitedBDXMessageHandler.mm

```diff
@@ -0,0 +1,129 @@
+using namespace chip::Messaging;
+using namespace chip::bdx;
+
+MTROTAUnsolicitedBDXMessageHandler * MTROTAUnsolicitedBDXMessageHandler::sInstance = nullptr;
+
+CHIP_ERROR MTROTAUnsolicitedBDXMessageHandler::Init(System::Layer * systemLayer, ExchangeManager * exchangeManager)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturnError(systemLayer != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(exchangeManager != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    mSystemLayer = systemLayer;
+    mExchangeMgr = exchangeManager;
+    return mExchangeMgr->RegisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id, this);
+}
+
+MTROTAUnsolicitedBDXMessageHandler * MTROTAUnsolicitedBDXMessageHandler::GetInstance()
+{
+    return sInstance;
+}
+
+void MTROTAUnsolicitedBDXMessageHandler::Shutdown()
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturn(mExchangeMgr != nullptr);
+    mExchangeMgr->UnregisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id);
+
+    VerifyOrReturn(mOTAImageTransferHandler != nullptr);
+
+    delete mOTAImageTransferHandler;
+    mOTAImageTransferHandler = nullptr;
+}
+
+void MTROTAUnsolicitedBDXMessageHandler::ControllerShuttingDown(MTRDeviceController_Concrete * controller)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturn(mOTAImageTransferHandler != nullptr);
+
+    FabricIndex peerFabricIndex = mOTAImageTransferHandler->GetPeerFabricIndex();
+    VerifyOrReturn(peerFabricIndex != kUndefinedFabricIndex);
+
+    if (controller.fabricIndex == peerFabricIndex)
+    {
+        delete mOTAImageTransferHandler;
+        mOTAImageTransferHandler = nullptr;
+    }
+}
+
+bool MTROTAUnsolicitedBDXMessageHandler::IsInAnOngoingTransfer()
+{
+    return (mOTAImageTransferHandler != nullptr);
+}
+
+CHIP_ERROR MTROTAUnsolicitedBDXMessageHandler::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, const SessionHandle & session,
+    ExchangeDelegate * _Nonnull & newDelegate)
+{
+    assertChipStackLockedByCurrentThread();
+
+    ChipLogDetail(BDX, "MTROTAUnsolicitedBDXMessageHandler: OnUnsolicitedMessageReceived: message " ChipLogFormatMessageType " protocol " ChipLogFormatProtocolId,
+        payloadHeader.GetMessageType(), ChipLogValueProtocolId(payloadHeader.GetProtocolID()));
+
+    VerifyOrReturnError(mExchangeMgr != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    // Only proceed if there is a valid fabric index for the SessionHandle.
+    if (session->IsSecureSession() && session->AsSecureSession() != nullptr && session->AsSecureSession()->GetFabricIndex() != kUndefinedFabricIndex) {
+
+        // If we receive a ReceiveInit BDX message, create a new MTROTAImageTransferHandler and register it
+        // as the handler for all BDX messages that will come over this exchange.
+        if (payloadHeader.HasMessageType(MessageType::ReceiveInit)) {
+            mOTAImageTransferHandler = new MTROTAImageTransferHandler(mSystemLayer);
```

Comment:
So... this part is wrong.  If `mOTAImageTransferHandler` is already not null, we will clobber the pointer and then things will go very badly wrong on shutdown.  The fact that we apparently have no tests exercising this case is very bad and we need to fix that.

We need to do one of two things:

1. _Actually support multiple instances of MTROTAImageTransferHandler being live_, and then we can create a new one here, or:
2. Have a check for `IsInAnOngoingTransfer()` and fail out if it's true here.

Note that this is for the "we got an unsolicited BDX message" case.  The reason our tests work is that we _do_ respond BUSY from the QueryImage command if we are in a transfer, which would be _before_ the other side tries to start BDX, so in "normal" operation we never reach this code with IsInAnOngoingTransfer() true.

So long story short, for now

* Put a IsInAnOngoingTransfer() check at the top of this method and error out if it's true.
* Do not set mOTAImageTransferHandler here.  It gets set in OnTransferHandlerCreated.
* As a followup, we start supporting multiple MTROTAImageTransferHandler being live.

Reply:
Added to follow up - https://github.com/project-chip/connectedhomeip/issues/36526